### PR TITLE
Pr1.1/reduce status file error

### PR DESCRIPTION
**Pull request by feng-j678**
[X] add new error details for compressed assessment
[X] add new error details for compressed installation
[x] set error object code to warning if file is truncated
[x] unit tests - remove the global var expected_truncated_count using a dynamic var that will handle different environment run i.e running the test locally might yield different count than in the log because the error message contains log file path that's vary from env to env
[x] add new error message noting the packages have been truncated. 
[x] during transition state, the status will not changed to warning, even the file is truncated, but it will contain the truncated error message
[x] current logic is status file mirror complete status file
No truncation:
status file size is same as complete status, no truncation message, status no change
Yes truncation:
status file size < complete status size, yes truncation message and status to warning
terminal state, if complete status file happens < 126kb, the status file will be equal to complete status file

[x]Note: this pr merged modified the __set_errors_json logic https://github.com/Azure/LinuxPatchExtension/pull/223/files#diff-ba04b78bd4fbe163e8379c82e4a534f5ced5fb2ebdd5f94b00cbcfe45eb8de77
when there's informational error message, it will always reset the errors_by_operation and errors_count_by_operation, so I move the truncated message logic into a separate method __recompose_truncated_substatus_msg_errors that way the reset will not remove the truncated message and keep the logic separately as it writes to the status file

### CODE REVIEW

File: src/core/src/service_interfaces/StatusHandler.py [hunk 1]
```diff
@@ -798,17 +836,209 @@ def __try_add_error(error_list, detail):
...
-        message += " The latest {0} error/s are shared in detail. To view all errors, review this log file on the machine: {1}".format(len(errors_by_operation), self.__log_file_path) if error_count_by_operation > 0 else ""
+        success_code = Constants.PatchOperationTopLevelErrorCode.SUCCESS
+        error_code = Constants.PatchOperationTopLevelErrorCode.ERROR
+        error_count = error_count_by_operation
+        code = success_code if error_count_by_operation == 0 else error_code
+
+        # Update the errors json to include truncation detail
+        if truncated:
+            error_count += 1    # add 1 because of truncation
+            code = Constants.PatchOperationTopLevelErrorCode.WARNING if code != error_code else error_code
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:849]: General comment for all code - keep it simple and avoid unnecessary variable assignment where not justified. This code for example, can just be condensed into 3 succinct lines from 9:

error_count_by_operation += 1 if truncated else 0
code = Constants.PatchOperationTopLevelErrorCode.SUCCESS if error_count_by_operation == 0 else Constants.PatchOperationTopLevelErrorCode.ERROR
code = Constants.PatchOperationTopLevelErrorCode.WARNING if (truncated and error_count_by_operation == 1) else code    # the only entry relates to the truncation

Too many lines of code that don't justify existence through necessary clarity or in promoting code-reuse must be avoided. They increase long-term maintainability overhead significantly when present across the code base. Please follow this ask in all PRs and in other places in this PR.

Comment (kjohn-msft) [reply src/core/src/service_interfaces/StatusHandler.py:849]: @rane-rajasi to also review code with this lens

**Review by kjohn-msft — CHANGES_REQUESTED** [review]
Generic comment that applies to all code / this+other PRs inline. Only a very high-level pass. More comments may follow later when I free up.

File: src/core/src/service_interfaces/StatusHandler.py [hunk 2]
```diff
@@ -817,17 +817,24 @@ def __try_add_error(error_list, detail):
...
+
         if error_count_by_operation == 1 and errors_by_operation[0]['code'] == Constants.PatchOperationErrorCodes.INFORMATIONAL:    # special-casing for single informational messages
             message = errors_by_operation[0]['message']
             errors_by_operation = []
-            error_count_by_operation = 0
         else:
+            # Update msg error code to warning for truncation
+            if is_status_truncated:
+                error_count_by_operation += 1  # add 1 because of truncation creates a new error detail object
+                code = Constants.PatchOperationTopLevelErrorCode.WARNING if code != Constants.PatchOperationTopLevelErrorCode.ERROR else Constants.PatchOperationTopLevelErrorCode.ERROR
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:831]: Test this on a machine where the results go into CRP API / ARG.

File: src/core/src/service_interfaces/StatusHandler.py [hunk 3]
```diff
@@ -1019,24 +1028,37 @@ def __get_substatus_index(self, substatus_list_name, substatus_list):
...
         self.composite_logger.log_verbose("Recompose truncated substatus")
-        truncated_substatus_message = self.__update_patches_in_substatus(substatus_msg=substatus_message, substatus_msg_patches=truncated_patches)
+        truncated_substatus_message = self.__update_patches_in_substatus(substatus_msg=substatus_message, substatus_msg_patches=truncated_patches, substatus_msg_errors=truncated_msg_errors)
 
         truncated_status_file['status']['substatus'][substatus_index]['formattedMessage']['message'] = json.dumps(truncated_substatus_message)
         return truncated_status_file
 
-    def __update_patches_in_substatus(self, substatus_msg, substatus_msg_patches):
-        """ update the substatus message patches """
+    def __recompose_substatus_msg_errors(self, errors_details_list, count_total_errors):
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:1048]: Rename to __recompose_truncated_substatus_msg_errors

File: src/core/src/service_interfaces/StatusHandler.py [hunk 4]
```diff
@@ -1019,24 +1028,37 @@ def __get_substatus_index(self, substatus_list_name, substatus_list):
...
+        """ Recompose truncated substatus message errors json """
+        truncated_error_detail = self.__set_error_detail(Constants.PatchOperationErrorCodes.TRUNCATION, Constants.StatusTruncationConfig.TRUNCATION_WARNING_MESSAGE)  # Reuse the errors object set up
+        self.__try_add_error(errors_details_list, truncated_error_detail)  # add new truncated error detail to beginning in errors details list
+        truncated_errors_json = self.__set_errors_json(count_total_errors, errors_details_list, is_status_truncated=True)
+
+        return truncated_errors_json
+
+    def __update_patches_in_substatus(self, substatus_msg, substatus_msg_patches, substatus_msg_errors=None):
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:1056]: rename to __update_patches_and_errors_in_substatus

File: src/core/tests/Test_StatusHandlerTruncation.py [hunk 5]
```diff
@@ -414,26 +401,28 @@ def test_both_assessment_and_installation_with_status_error_truncated(self):
...
 
     def test_truncation_method_time_performance(self):
-        """ Comparing truncation code performance on prior and post on 1000 packages with frequency of 300
-        assert truncation code logic time performance is 30 secs more than current (prior truncation) logic"""
+        """ Comparing truncation code performance on prior and post on 750 packages with frequency of 30
+        assert truncation code logic time performance is 30 secs more than current (prior truncation) logic """
+        start_index = 0
+        end_index = 30
+        patch_count = 350
+        expected_time_performance = 30
```
Comment (kjohn-msft) [inline src/core/tests/Test_StatusHandlerTruncation.py:416]: TimePerformanceConfig._blah_

**Review by kjohn-msft — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/src/service_interfaces/StatusHandler.py (hunk 2 above)
Comment (feng-j678) [reply src/core/src/service_interfaces/StatusHandler.py:831]: the log from condition check has this logic, so far no error from it

done

File: src/core/src/service_interfaces/StatusHandler.py (hunk 3 above)
Comment (feng-j678) [reply src/core/src/service_interfaces/StatusHandler.py]: done

File: src/core/src/service_interfaces/StatusHandler.py (hunk 4 above)
Comment (feng-j678) [reply src/core/src/service_interfaces/StatusHandler.py]: done

File: src/core/tests/Test_StatusHandlerTruncation.py (hunk 5 above)
Comment (feng-j678) [reply src/core/tests/Test_StatusHandlerTruncation.py]: file is created in /tests/library

done?

File: src/core/src/service_interfaces/StatusHandler.py [hunk 6]
```diff
@@ -817,17 +817,24 @@ def __try_add_error(error_list, detail):
...
 
-    def __set_errors_json(self, error_count_by_operation, errors_by_operation):
+    def __set_errors_json(self, error_count_by_operation, errors_by_operation, is_status_truncated=False):
         """ Compose the error object json to be added in 'errors' in given operation's summary """
+        code = Constants.PatchOperationTopLevelErrorCode.SUCCESS if error_count_by_operation == 0 else Constants.PatchOperationTopLevelErrorCode.ERROR
+
         if error_count_by_operation == 1 and errors_by_operation[0]['code'] == Constants.PatchOperationErrorCodes.INFORMATIONAL:    # special-casing for single informational messages
             message = errors_by_operation[0]['message']
             errors_by_operation = []
-            error_count_by_operation = 0
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:825]: Why is this changing?

File: src/core/src/service_interfaces/StatusHandler.py [hunk 7]
```diff
@@ -1004,39 +1013,52 @@ def __size_of_constant_status_data(self, status_payload_json, assessment_status_
...
 
         if installation_status_index is not None:
-            installation_msg_without_patches = self.__update_patches_in_substatus(substatus_msg=self.__installation_substatus_msg_copy, substatus_msg_patches=[])
+            installation_msg_without_patches = self.__update_patches_and_errors_in_substatus(substatus_msg=self.__installation_substatus_msg_copy, substatus_msg_patches=[])
             status_file_no_list_data['status']['substatus'][installation_status_index]['formattedMessage']['message'] = json.dumps(installation_msg_without_patches)
         return self.__calc_status_size_on_disk(json.dumps(status_file_no_list_data))
 
-    def __get_substatus_index(self, substatus_list_name, substatus_list):
+    def __get_substatus_index_and_status(self, substatus_list_name, substatus_list):
         """ Gets the index of an operation substatus in the overall extension status for further operations """
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:1025]: nit: edit the description as per new changes

File: src/core/tests/Test_StatusHandlerTruncation.py [hunk 8]
```diff
@@ -74,18 +72,17 @@ def test_only_assessment_patches_over_size_limit_truncated(self):
         """ Perform truncation on very large assessment patches and checks for time performance concern.
         Before truncation: 100000 assessment patches in status
         complete status file byte size: 19,022kb,
-        Expected (After truncation): 672 assessment patches in status
+        Expected (After truncation): ~671 assessment patches in status
         operation: Assessment,
         assessment substatus name: PatchAssessmentSummary,
         assessment substatus status: warning,
         assessment errors code: 0 (success),
```
Comment (rane-rajasi) [inline src/core/tests/Test_StatusHandlerTruncation.py:79]: this should change to 2 (warning)

Applicable in other tests too

File: src/core/tests/library/TruncationTimePerformanceConfig.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+# Requires Python 2.7+
+
+
+class TruncationTimePerformanceConfig(object):
+    """ Helps set up constants for comparing time performance of prior and post truncation logic unit test """
+    class UTConfig:
+        MIN_OPERATION_ITERATIONS = 0
+        MAX_OPERATION_ITERATIONS = 30
+        NUMBER_OF_PATCHES = 350
+        EXPECTED_TRUNCATION_TIME_LIMIT_IN_SEC = 30
```
Comment (rane-rajasi) [inline src/core/tests/library/TruncationTimePerformanceConfig.py:24]: I think the idea was to move all time performance related tests in here. Please check with Koshy once

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/src/service_interfaces/StatusHandler.py (hunk 6 above)
Comment (feng-j678) [reply src/core/src/service_interfaces/StatusHandler.py:825]: i reverted the changes back to what it was before, so i created a new set_truncated_errors method to separate the logic becuase currently logic reset the details[] which wipe out the truncation details

File: src/core/src/service_interfaces/StatusHandler.py [hunk 10]
```diff
@@ -817,34 +817,32 @@ def __try_add_error(error_list, detail):
...
 
         return {
-            "code": code,
+            "code": Constants.PatchOperationTopLevelErrorCode.SUCCESS if error_count_by_operation == 0 else Constants.PatchOperationTopLevelErrorCode.ERROR,
             "details": errors_by_operation,
             "message": message
         }
 
-    def __set_error_detail(self, error_code, message):
+    def __set_errors_detail(self, error_code, message):
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:835]: _set_error_detail

**Review by kjohn-msft — APPROVED** [review]

**Pull request merged**